WIP on disk monitor leaks

yuzefovich opened the pull request:
Epic: None

yuzefovich committed:
WIP on stopping temp storage disk monitor

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/120599)
<!-- Reviewable:end -->

yuzefovich committed:
flowinfra: always call ConsumerClosed on flow cleanup

This commit adjusts the flow cleanup to always call
`execinfra.RowSource.ConsumerClosed` method unconditionally. Generally,
this is not necessary (because the row source is exhausted or
ConsumerClosed is called in the `execinfra.Run` loop), but in edge cases
the method might never be called (e.g. when the row source is created
but is never `Run` due to server shutting down). As a result, some
resources (like memory monitors) might not be released in such edge
cases, and even though it's not a problem on its own, it will be when
we introduce some assertions on the server shutdown around memory
monitors.

This commit adjusts the contract of `ConsumerClosed` method to explicity
call out that multiple calls are ok and fixes a single instance where it
wasn't.

Release note: None

yuzefovich committed:
sql: simplify newFlowCtxForExplainPurposes a bit

This commit makes it so that we don't use the planner's monitor as
a parent for the monitor that is created for EXPLAIN (VEC) which allows
us to remove some cleanup (and makes introducing some assertions
easier).

Release note: None

yuzefovich committed:
util/mon: assert that short-living monitors are stopped

To be filled.

Release note: None

yuzefovich committed:
WIP on CDC fixes

WIP on cdc tests

WIP on marking CDC flows as long-living

There is a race between shutting down goroutines running the CDC flow
(which would close a few memory monitors) and closing the connExecutor
handling the connection (which closes the "txn" monitor which is an
ancestor of the flow monitors).

yuzefovich commented:
Superseded by #119794
